### PR TITLE
[AMDGPU] Fix lowering of abs for i16 vectors with more than 2 elements

### LINKED ISSUE #94606: AMDGPU lowering of abs is bad for i16 vectors with more than 2 elements
We currently get a simple 2-instruction expansion in the v2i16 case, but larger vectors scalarize and produce bad code instead of vector splitting 

```
; RUN: llc -march=amdgcn -mcpu=gfx900 < %s

; v_pk_sub_i16 v1, 0, v0
; v_pk_max_i16 v0, v0, v1
define <2 x i16> @v_abs_v2i16(<2 x i16> %arg) {
  %res = call <2 x i16> @llvm.abs.v2i16(<2 x i16> %arg, i1 false)
  ret <2 x i16> %res
}

; This should decompose into 2 x v2i16 and 4 instructions, but we get this long sequence

;	v_lshrrev_b32_e32 v2, 16, v1
;	v_sub_u16_e32 v3, 0, v2
;	v_max_i16_e32 v2, v2, v3
;	v_lshrrev_b32_e32 v3, 16, v0
;	v_sub_u16_e32 v4, 0, v3
;	v_max_i16_e32 v3, v3, v4
;	v_sub_u16_e32 v4, 0, v1
;	v_max_i16_e32 v1, v1, v4
;	v_sub_u16_e32 v4, 0, v0
;	v_max_i16_e32 v0, v0, v4
;	s_mov_b32 s4, 0x5040100
;	v_perm_b32 v0, v3, v0, s4
;	v_perm_b32 v1, v2, v1, s4
define <4 x i16> @v_abs_v4i16(<4 x i16> %arg) {
  %res = call <4 x i16> @llvm.abs.v4i16(<4 x i16> %arg, i1 false)
  ret <4 x i16> %res
}
```

### PR DESCRIPTION
**Pull request by tgymnich**
fixes #94606

Expansion of `ABS` for `i16` vectors with more than 2 elements is currently falling back to scalarization of the vector.
This PR adds a custom lowering for `ABS` on `i16` vectors that splits the vector into multiple `<2 x i 16>` vectors.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/abs.ll [hunk 1]
```diff
@@ -0,0 +1,1152 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx600 < %s | FileCheck -check-prefix=GFX6 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx700 < %s | FileCheck -check-prefix=GFX7 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx803 < %s | FileCheck -check-prefix=GFX8 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx900 < %s | FileCheck -check-prefix=GFX9 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx1010 < %s | FileCheck -check-prefix=GFX10 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx1100 < %s | FileCheck -check-prefix=GFX11 %s
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx1200 < %s | FileCheck -check-prefix=GFX12 %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/abs.ll:9]: We apparently don't have existing tests for this, so you should also test the scalar, 2 and 3 element cases, and maybe 6.

Also we should have the 32/64cases, but I guess you can just add an i16 suffix to the test and leave those for later. Also would be nice to cover the basic SALU cases

Comment (tgymnich) [reply llvm/test/CodeGen/AMDGPU/abs.ll]: done

**Review by arsenm — APPROVED** [review]

**Comment by nico** [comment]
Looks like this breaks tests: http://45.33.8.238/linux/140660/step_11.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by jrbyrnes** [comment]
https://github.com/llvm/llvm-project/pull/95620

**Pull request merged**